File: Android/EasyBudget/app/src/main/java/com/benoitletondor/easybudgetapp/model/Expense.java
/*
 *   Copyright 2015 Benoit LETONDOR
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package com.benoitletondor.easybudgetapp.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.benoitletondor.easybudgetapp.helper.DateHelper;

import java.util.Date;

/**
 * Object that represent an expense.
 *
 * @author Benoit LETONDOR
 */
public class Expense implements Parcelable
{
    /**
     * DB id of this expense (can be null)
     */
    private Long id;
    /**
     * Title of the expense
     */
    private String  title;
    /**
     * Amount of this expense (can be < 0).
     */
    private double  amount;
    /**
     * Date of the expense
     */
    private Date    date;
    /**
     * Associated recurring expense (can be null if not a recurring one)
     */
    @Nullable
    private RecurringExpense recurringExpense;

// --------------------------------->

    /**
     *
     * @param title
     * @param amount
     * @param date
     */
    public Expense(@NonNull String title, double amount, @NonNull Date date)
    {
        this(null, title, amount, date, null);
    }

    /**
     *
     * @param title
     * @param amount
     * @param date
     * @param recurringExpense
     */
    public Expense(@NonNull String title, double amount, @NonNull Date date, @Nullable RecurringExpense recurringExpense)
    {
        this(null, title, amount, date, recurringExpense);
    }

    /**
     *
     * @param id
     * @param title
     * @param amount
     * @param date
     * @param recurringExpense
     */
    public Expense(Long id, @NonNull String title, double amount, @NonNull Date date, @Nullable RecurringExpense recurringExpense)
    {
        this.id = id;

        if ( title.isEmpty() )
        {
            throw new IllegalArgumentException("title is empty or null");
        }

        this.title = title;

        if( amount == 0 )
        {
            throw new IllegalArgumentException("amount should be != 0");
        }

        this.amount = amount;
        this.date = DateHelper.cleanDate(date);
        this.recurringExpense = recurringExpense;
    }

    /**
     *
     * @param in
     */
    private Expense(Parcel in)
    {
        id = (Long) in.readValue(Long.class.getClassLoader());
        title = in.readString();
        amount = in.readDouble();
        date = new Date(in.readLong());
        recurringExpense = in.readParcelable(RecurringExpense.class.getClassLoader());
    }

// --------------------------------->

    public Long getId()
    {
        return id;
    }

    public void setId(Long id)
    {
        this.id = id;
    }

    @Nullable
    public RecurringExpense getAssociatedRecurringExpense()
    {
        return recurringExpense;
    }

    public void setAssociatedRecurringExpense(@Nullable RecurringExpense recurringExpense)
    {
        this.recurringExpense = recurringExpense;
    }

    public boolean isRecurring()
    {
        return recurringExpense != null;
    }

    @NonNull
    public String getTitle()
    {
        return title;
    }

    public void setTitle(@NonNull String title)
    {
        this.title = title;
    }

    @NonNull
    public Date getDate()
    {
        return date;
    }

    public void setDate(@NonNull Date date)
    {
        this.date = DateHelper.cleanDate(date);
    }

    public double getAmount()
    {
        return amount;
    }

    public void setAmount(double amount)
    {
        this.amount = amount;
    }

    public boolean isRevenue()
    {
        return amount < 0;
    }

// --------------------------------->

    @Override
    public void writeToParcel(Parcel dest, int flags)
    {
        dest.writeValue(id);
        dest.writeString(title);
        dest.writeDouble(amount);
        dest.writeLong(date.getTime());
        dest.writeParcelable(recurringExpense, flags);
    }

    @Override
    public int describeContents()
    {
        return 0;
    }

    public static final Creator<Expense> CREATOR = new Creator<Expense>()
    {
        @Override
        public Expense createFromParcel(Parcel in)
        {
            return new Expense(in);
        }

        @Override
        public Expense[] newArray(int size)
        {
            return new Expense[size];
        }
    };
}

File: Android/EasyBudget/app/src/main/java/com/benoitletondor/easybudgetapp/model/Expense.kt
/*
 *   Copyright 2024 Benoit LETONDOR
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package com.benoitletondor.easybudgetapp.model

import android.os.Parcel
import android.os.Parcelable
import androidx.compose.runtime.Immutable
import java.time.LocalDate

@Immutable
data class Expense(val id: Long?,
                   val title: String,
                   val amount: Double,
                   val date: LocalDate,
                   val checked: Boolean,
                   val associatedRecurringExpense: AssociatedRecurringExpense?) : Parcelable {

    constructor(title: String,
                amount: Double,
                date: LocalDate,
                checked: Boolean) : this(null, title, amount, date, checked, null)

    constructor(id: Long,
                title: String,
                amount: Double,
                date: LocalDate,
                checked: Boolean) : this(id, title, amount, date, checked, null)

    constructor(title: String,
                amount: Double,
                date: LocalDate,
                checked: Boolean,
                associatedRecurringExpense: AssociatedRecurringExpense) : this(null, title, amount, date, checked, associatedRecurringExpense)

    private constructor(parcel: Parcel) : this(
        parcel.readValue(Long::class.java.classLoader) as? Long,
        parcel.readString()!!,
        parcel.readDouble(),
        LocalDate.ofEpochDay(parcel.readLong()),
        parcel.readInt() == 1,
        parcel.readParcelable(AssociatedRecurringExpense::class.java.classLoader)
    )

    init {
        if( title.isEmpty() ) {
            throw IllegalArgumentException("title is empty")
        }
    }

    fun isRevenue() = amount < 0

    fun isRecurring() = associatedRecurringExpense != null

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeValue(id)
        parcel.writeString(title)
        parcel.writeDouble(amount)
        parcel.writeLong(date.toEpochDay())
        parcel.writeInt(if( checked ) { 1 } else { 0 })
        parcel.writeParcelable(associatedRecurringExpense, flags)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<Expense> {
        override fun createFromParcel(parcel: Parcel): Expense {
            return Expense(parcel)
        }

        override fun newArray(size: Int): Array<Expense?> {
            return arrayOfNulls(size)
        }
    }

}